<a href="https://colab.research.google.com/github/alexlopespereira/mba_enap/blob/main/Exercicios/Aula9_Exercicio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aula 9 - Exercício - Agregação com Pandas

In [2]:
!pip install --upgrade ipynb_autograde > /dev/null

In [3]:
%env  COURSE=mba_enap
from ipynb_autograde.main import init_log, validate
init_log()
email_result=!gcloud config get-value account
if not email_result or 'unset' in email_result[0]:
  !gcloud auth login

env: COURSE=mba_enap
Go to the following link in your browser, and complete the sign-in prompts:

    https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=32555940559.apps.googleusercontent.com&redirect_uri=https%3A%2F%2Fsdk.cloud.google.com%2Fauthcode.html&scope=openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fappengine.admin+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fsqlservice.login+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcompute+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Faccounts.reauth&state=f0EmNIAWVdmGKz7mnXFeIaIzbjY6If&prompt=consent&token_usage=remote&access_type=offline&code_challenge=x-2rzrWSl_497aBIi7_VpFdHuxHJBMDJ5oU3GIIbKK4&code_challenge_method=S256

Once finished, enter the verification code provided in your browser: 4/0AanRRrvW9mPaZ3u4jQfXT1UBL8LombENRaleJeHaYWIIEfIDm8qoPuzNfwGyMXaVyr33xw

You are now logged in as [alexlopespereira@gmail.c

In [5]:
# Faca o import das bibliotecas
import pandas as pd
import numpy as np

### 9.1 Escreva um prompt de uma função python para carregar os dados de população de 2017 e realizar as operações ilustradas na tabela a seguir. Retorne apenas os top 3 resultados ordenados pela soma.

#### Entrada: ```https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/populacao/estimativa_dou_2017.xlsx```


### Resultado esperado:

|    | uf   |      sum |     mean |   median |    std |
|---:|:-----|---------:|---------:|---------:|-------:|
|  0 | SP   | 45094866 |  69914.5 |    13810 | 489370 |
|  1 | MG   | 21119536 |  24759.1 |     8287 | 100604 |
|  2 | RJ   | 16718956 | 181728   |    35918 | 693337 |


In [73]:
prompt91 = """Escreva uma função Python utilizando pandas para carregar um arquivo Excel de uma URL passada como 1o argumento interpretando a coluna cod_ibge como string. Nomeie esse dataframe como df_pop. Agrege o dataframe  por uf e calcule as métricas sum, mean, median e std da coluna pop2017, arredonde o resultado de todas essas métricas para uma casa decimal. Ordene o dataframe agregado pela coluna sum, e depois retorne os top 3 registros. Antes de retornar, aplique o metodo reset_index(drop=False)"""

In [74]:
validate(prompt91, "A9-E1")

Codigo gerado: 
{'code': "def process_excel_data(url):\n    df_pop = pd.read_excel(url, dtype={'cod_ibge': str})\n    aggregated_df = df_pop.groupby('uf')['pop2017'].agg(['sum', 'mean', 'median', 'std']).round(1)\n    sorted_df = aggregated_df.sort_values(by='sum', ascending=False)\n    top_3 = sorted_df.head(3).reset_index(drop=False)\n    return top_3", 'status': 'success', 'summary': {'passed_tests': 1, 'success_rate': '100.00%', 'total_tests': 1}, 'test_results': [{'actual': {'aggregation_summary': {'sums': {'mean': 276401.4, 'median': 58015.0, 'std': 1283310.7, 'sum': 82933358.0}, 'total_rows': {'max': 3, 'min': 3}}, 'column_analysis': {'expected_columns': ['uf', 'sum', 'mean', 'median', 'std'], 'missing_columns': [], 'possible_matches': {}, 'provided_columns': ['uf', 'sum', 'mean', 'median', 'std'], 'unexpected_columns': []}, 'verified_samples': [{'expected_values': {'sum': 45094866}, 'filter': {'uf': 'SP'}}, {'expected_values': {'sum': 16718956}, 'filter': {'uf': 'RJ'}}]}, 'erro

In [58]:
# prompt: Escreva uma função Python utilizando pandas para carregar um arquivo Excel de uma URL passada como 1o argumento interpretando a coluna cod_ibge como string. Nomeie esse dataframe como df_pop. Agrege o dataframe  por uf e calcule as métricas sum, mean, median e std da coluna pop2017, arredonde o resultado de todas essas métricas para uma casa decimal. Ordene o dataframe agregado pela coluna sum, e depois retorne os top 3 registros. Antes de retornar, aplique o metodo reset_index(drop=False)

import pandas as pd
import numpy as np

def aggregate_population_data(excel_url):
    try:
        df_pop = pd.read_excel(excel_url, dtype={'cod_ibge': str})
        df_pop_grouped = df_pop.groupby('uf')['pop2017'].agg(['sum', 'mean', 'median', 'std'])
        df_pop_grouped = df_pop_grouped.round(1)
        df_pop_grouped = df_pop_grouped.sort_values(by='sum', ascending=False)
        df_pop_grouped = df_pop_grouped.head(3).reset_index(drop=False)
        return df_pop_grouped
    except Exception as e:
        print(f"An error occurred: {e}")
        return None

In [60]:
df = aggregate_population_data('https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/populacao/estimativa_dou_2017.xlsx')
df.to_markdown()

'|    | uf   |      sum |     mean |   median |    std |\n|---:|:-----|---------:|---------:|---------:|-------:|\n|  0 | SP   | 45094866 |  69914.5 |    13810 | 489370 |\n|  1 | MG   | 21119536 |  24759.1 |     8287 | 100604 |\n|  2 | RJ   | 16718956 | 181728   |    35918 | 693337 |'

In [14]:
df.to_markdown()

'| uf   |         sum |     mean |   median |    std |\n|:-----|------------:|---------:|---------:|-------:|\n| SP   | 4.50949e+07 |  69914.5 |    13810 | 489370 |\n| MG   | 2.11195e+07 |  24759.1 |     8287 | 100604 |\n| RJ   | 1.6719e+07  | 181728   |    35918 | 693337 |'

### 9.2 Escreva um prompt de uma função python para carregar os dados do dataset do Titanic e realizar as operações ilustradas na tabela a seguir.

#### Entrada: ```https://github.com/alexlopespereira/mba_enap/raw/main/data/originais/titanic/titanic.xls```


### Resultado esperado:


|    |   pclass | sex    |   survived |   fare |
|---:|---------:|:-------|-----------:|-------:|
|  0 |        1 | female |        139 |   80.9 |
|  1 |        1 | male   |         61 |   49.5 |
|  2 |        2 | female |         94 |   23   |
|  3 |        2 | male   |         25 |   13   |
|  4 |        3 | female |        106 |   10.5 |
|  5 |        3 | male   |         75 |    7.9 |


In [75]:
prompt92 = """Escreva uma função Python utilizando pandas para carregar um arquivo Excel de uma URL passada como 1o argumento. Nomeie esse dataframe como df_titanic. Agrege o dataframe por pclass e sex, e calcule a soma da coluna survived e a mediana da coluna fare, assim: df_titanic.groupby(['pclass', 'sex']).agg({'survived': 'sum', 'fare': 'median'}). Arredonde o resultado de todas essas métricas para uma casa decimal. Antes de retornar, aplique o metodo reset_index(drop=False)"""

In [76]:
validate(prompt92, "A9-E2")

Codigo gerado: 
{'code': "def process_titanic_data(url):\n    df_titanic = pd.read_excel(url)\n    aggregated_df = df_titanic.groupby(['pclass', 'sex']).agg({'survived': 'sum', 'fare': 'median'})\n    aggregated_df = aggregated_df.round(1)\n    result_df = aggregated_df.reset_index(drop=False)\n    return result_df", 'status': 'success', 'summary': {'passed_tests': 1, 'success_rate': '100.00%', 'total_tests': 1}, 'test_results': [{'actual': {'aggregation_summary': {'sums': {'fare': 184.8, 'pclass': 12.0, 'survived': 500.0}, 'total_rows': {'max': 6, 'min': 6}}, 'column_analysis': {'expected_columns': ['pclass', 'sex', 'survived', 'fare'], 'missing_columns': [], 'possible_matches': {}, 'provided_columns': ['pclass', 'sex', 'survived', 'fare'], 'unexpected_columns': []}, 'verified_samples': [{'expected_values': {'survived': 139}, 'filter': {'pclass': 1, 'sex': 'female'}}, {'expected_values': {'survived': 75}, 'filter': {'pclass': 3, 'sex': 'male'}}]}, 'error': [], 'expected': {'data': {'a

In [61]:
# prompt: Escreva uma função Python utilizando pandas para carregar um arquivo Excel de uma URL passada como 1o argumento. Nomeie esse dataframe como df_titanic. Agrege o dataframe por pclass e sex, e calcule a soma da coluna survived e a mediana da coluna fare, assim: df_titanic.groupby(['pclass', 'sex']).agg({'survived': 'sum', 'fare': 'median'}). Arredonde o resultado de todas essas métricas para uma casa decimal. Antes de retornar, aplique o metodo reset_index(drop=False)

def aggregate_titanic_data(excel_url):
    try:
        df_titanic = pd.read_excel(excel_url)
        df_titanic_grouped = df_titanic.groupby(['pclass', 'sex']).agg({'survived': 'sum', 'fare': 'median'})
        df_titanic_grouped = df_titanic_grouped.round(1)
        df_titanic_grouped = df_titanic_grouped.reset_index(drop=False)
        return df_titanic_grouped
    except Exception as e:
        print(f"An error occurred: {e}")
        return None

In [66]:
aggregate_titanic_data("https://github.com/alexlopespereira/mba_enap/raw/main/data/originais/titanic/titanic.xls")

pclass     sex  survived  fare
0       1  female       139  80.9
1       1    male        61  49.5
2       2  female        94  23.0
3       2    male        25  13.0
4       3  female       106  10.5
5       3    male        75   7.9

### 9.3 Escreva uma função para realizar o agrupamento do dataset de morbidade por UF e por município aplicando a soma sobre as colunas Valor_total, Internações e Dias_permanência. Em seguida calcule, para todos os municípios nesta agregação, o custo médio diário das internações com a fórmula abaixo. Retorne um dataframe com os top 3 municipios segundo o custo médio de internação.

#### Entrada: ```https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv```


#### Resultado esperado:


|      |   uf |   cod_ibge6 | Município             |      Valor_total |   Internações |   Dias_permanência |   custo_medio_diario_intern |
|-----:|-----:|------------:|:----------------------|-----------------:|--------------:|-------------------:|----------------------------:|
| 3967 |   41 |      410400 | Campina Grande do Sul |      3.32947e+07 |          7037 |              29277 |                      1137.2 |
| 3864 |   35 |      355365 | Taquaral              | 382405           |           152 |                465 |                       822.4 |
| 3685 |   35 |      353715 | Pedrinhas Paulista    | 769964           |           188 |               1004 |                       766.9 |
| 3205 |   33 |      330205 | Italva                |      2.39501e+06 |           585 |               3160 |                       757.9 |
| 3193 |   33 |      330115 | Cardoso Moreira       |      2.80194e+06 |           740 |               3840 |                       729.7 |

In [82]:
prompt93 = """A função tem no 1o argumento o caminho de um arquivo csv carregue o arquivo corretamente num dataframe chamado df_morbidade. Nesse csv, o caracter do argumento sep é o ";". Os na values são ['-', '...'], o simbolo para decimal é a virgula, e a coluna cod_ibge6 deve ser importada como string. A primeira coluna já tem o índice para o dataframe, ou seja, use index_col=0. Crie uma coluna chamada uf apartir dos dois primeiros caracteres da coluna cod_ibge6. Agrege por uf, cod_ibge6 e Município e calcule a soma do Valor_total, Internações e Dias_permanência.  Arredonde o resultado de todas essas métricas para uma casa decimal. Use a seguinte fórmula para calculár o custo diário médio de internação: custo_medio_diario_intern = Valor_total/Dias_permanência, e armazene esse resultado numa nova coluna do dataframe. Retorne os 5 registros de maior custo medio de internação. Antes de retornar, aplique o metodo reset_index(drop=False)"""

In [83]:
validate(prompt93, "A9-E3")

Codigo gerado: 
{'code': "def processar_dados_morbidade(caminho_arquivo, sep):\n    df_morbidade = pd.read_csv(caminho_arquivo, sep=sep, na_values=['-', '...'], decimal=',', \n                                dtype={'cod_ibge6': str}, index_col=0)\n    \n    df_morbidade['uf'] = df_morbidade['cod_ibge6'].str[:2]\n    \n    df_aggregated = df_morbidade.groupby(['uf', 'cod_ibge6', 'Município']).agg(\n        Valor_total=('Valor_total', 'sum'),\n        Internações=('Internações', 'sum'),\n        Dias_permanência=('Dias_permanência', 'sum')\n    ).round(1).reset_index(drop=False)\n    \n    df_aggregated['custo_medio_diario_intern'] = df_aggregated['Valor_total'] / df_aggregated['Dias_permanência']\n    \n    return df_aggregated.nlargest(5, 'custo_medio_diario_intern')", 'status': 'success', 'summary': {'passed_tests': 0, 'success_rate': '0.00%', 'total_tests': 1}, 'test_results': [{'actual': None, 'error': "Error while running your function: processar_dados_morbidade() missing 1 require

In [67]:
# prompt: A função tem no 1o argumento o caminho de um arquivo csv carregue o arquivo corretamente num dataframe chamado df_morbidade. Nesse csv, o caracter do argumento sep é o ";". Os na values são ['-', '...'], o simbolo para decimal é a virgula, e a coluna cod_ibge6 deve ser importada como string. A primeira coluna já tem o índice para o dataframe, ou seja, use index_col=0. Crie uma coluna chamada uf apartir dos dois primeiros caracteres da coluna cod_ibge6. Agrege por uf, cod_ibge6 e Município e calcule a soma do Valor_total, Internações e Dias_permanência.  Arredonde o resultado de todas essas métricas para uma casa decimal. Use a seguinte fórmula para calculár o custo diário médio de internação: custo_medio_diario_intern = Valor_total/Dias_permanência, e armazene esse resultado numa nova coluna do dataframe. Antes de retornar, aplique o metodo reset_index(drop=False). Retorne os 5 registros de maior custo medio de internação.  Escreva um código para testar esta função a partir desta url 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'

import pandas as pd
import numpy as np

def calcula_custo_medio_internacao(url):
    df_morbidade = pd.read_csv(url, sep=';', na_values=['-', '...'], decimal=',', dtype={'cod_ibge6': str}, index_col=0)
    df_morbidade['uf'] = df_morbidade['cod_ibge6'].str[:2]
    df_morbidade = df_morbidade.groupby(['uf', 'cod_ibge6', 'Município'])[['Valor_total', 'Internações', 'Dias_permanência']].sum().round(1)
    df_morbidade['custo_medio_diario_intern'] = (df_morbidade['Valor_total'] / df_morbidade['Dias_permanência']).round(1)
    df_morbidade = df_morbidade.reset_index(drop=False)
    df_morbidade = df_morbidade.sort_values('custo_medio_diario_intern', ascending=False).head(5)
    return df_morbidade

# teste
url = 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'
df_resultado = calcula_custo_medio_internacao(url)
df_resultado

uf cod_ibge6              Município  Valor_total  Internações  \
3967  41    410400  Campina Grande do Sul   33294700.7       7037.0   
3864  35    355365               Taquaral     382405.0        152.0   
3685  35    353715     Pedrinhas Paulista     769964.0        188.0   
3205  33    330205                 Italva    2395014.9        585.0   
3193  33    330115        Cardoso Moreira    2801938.8        740.0   

      Dias_permanência  custo_medio_diario_intern  
3967           29277.0                     1137.2  
3864             465.0                      822.4  
3685            1004.0                      766.9  
3205            3160.0                      757.9  
3193            3840.0                      729.7

In [81]:
def process_morbidade_data(file_path, sep=';'):
    df_morbidade = pd.read_csv(file_path, sep=sep, na_values=['-', '...'], decimal=',', index_col=0, dtype={'cod_ibge6': str})

    df_morbidade['uf'] = df_morbidade['cod_ibge6'].str[:2]

    aggregated_df = df_morbidade.groupby(['uf', 'cod_ibge6', 'Município']).agg(
        Valor_total=('Valor_total', 'sum'),
        Internações=('Internações', 'sum'),
        Dias_permanência=('Dias_permanência', 'sum')
    ).reset_index(drop=False)

    aggregated_df = aggregated_df.round(1)

    aggregated_df['custo_medio_diario_intern'] = aggregated_df['Valor_total'] / aggregated_df['Dias_permanência']

    top_costs = aggregated_df.nlargest(5, 'custo_medio_diario_intern')

    return top_costs.reset_index(drop=False)

url = 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'
df_resultado = process_morbidade_data(url)
df_resultado

index  uf cod_ibge6              Município  Valor_total  Internações  \
0   3967  41    410400  Campina Grande do Sul   33294700.7       7037.0   
1   3864  35    355365               Taquaral     382405.0        152.0   
2   3685  35    353715     Pedrinhas Paulista     769964.0        188.0   
3   3205  33    330205                 Italva    2395014.9        585.0   
4   3193  33    330115        Cardoso Moreira    2801938.8        740.0   

   Dias_permanência  custo_medio_diario_intern  
0           29277.0                1137.230614  
1             465.0                 822.376344  
2            1004.0                 766.896414  
3            3160.0                 757.916108  
4            3840.0                 729.671562

### 9.4 Crie um prompt de uma função python para retornar a soma do Valor_total dos 5 municípios (do exercício anterior) do dataset de Morbidade do Datasus (ao longo do período representado no dataset disponibilizado).

#### Entradas: ```https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv``` e ```["410400", "330115", "330205", "353715", "355365"]```

#### Resultado esperado:
|    | Município             |      Valor_total |
|---:|:----------------------|-----------------:|
|  0 | Campina Grande do Sul |      33294700.7 |
|  1 | Cardoso Moreira       |      2801938.8 |
|  2 | Italva                |      2395014.9 |
|  3 | Pedrinhas Paulista    | 769964.0           |
|  4 | Taquaral              | 382405.0           |



In [79]:
prompt94 = """A função tem no 1o argumento o caminho de um arquivo csv carregue o arquivo corretamente num dataframe chamado df_morbidade. Nesse csv, o caracter do argumento sep é o ";". Os na values são ['-', '...'], o simbolo para decimal é a virgula, e a coluna cod_ibge6 deve ser importada como string. A primeira coluna já tem o índice para o dataframe, ou seja, use index_col=0. Crie uma coluna chamada uf a partir dos dois primeiros caracteres da coluna cod_ibge6.  Selecione todos os registros cujo cod_ibge6 esteja nesta lista ["410400", "330115", "330205", "353715", "355365"]. Depois agregue por Município e retorne a soma do valor Valor_total. Escreva um código para testar esta função a partir desta url 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'. Antes de retornar, aplique o metodo reset_index(drop=False)"""

In [80]:
validate(prompt94, "A9-E4")

Error: 500 - Internal Server Error
Details: {"error": "Cloud function error: {\"message\":\"No test cases found for function_id: A9-E4\",\"status\":\"error\"}\n"}


In [71]:
# prompt: A função tem no 1o argumento o caminho de um arquivo csv carregue o arquivo corretamente num dataframe chamado df_morbidade. Nesse csv, o caracter do argumento sep é o ";". Os na values são ['-', '...'], o simbolo para decimal é a virgula, e a coluna cod_ibge6 deve ser importada como string. A primeira coluna já tem o índice para o dataframe, ou seja, use index_col=0. Crie uma coluna chamada uf a partir dos dois primeiros caracteres da coluna cod_ibge6.  Selecione todos os registros cujo cod_ibge6 esteja nesta lista ["410400", "330115", "330205", "353715", "355365"]. Depois agregue por Município e retorne a soma do valor Valor_total. Arredonde a soma do Valor_total com 1 casa decimal. Antes de retornar, aplique o metodo reset_index(drop=False). Escreva um código para testar esta função a partir desta url 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'.

def aggregate_morbidity_data(csv_url, cod_ibge_list):
    df_morbidade = pd.read_csv(csv_url, sep=';', na_values=['-', '...'], decimal=',', dtype={'cod_ibge6': str}, index_col=0)
    df_morbidade['uf'] = df_morbidade['cod_ibge6'].str[:2]
    df_morbidade = df_morbidade[df_morbidade['cod_ibge6'].isin(cod_ibge_list)]
    df_morbidade = df_morbidade.groupby('Município')['Valor_total'].sum().round(1).reset_index(drop=False)
    return df_morbidade

# Test
url = 'https://github.com/alexlopespereira/mba_enap/raw/refs/heads/main/data/originais/morbidade/consolidado/morbidade.csv'
cod_ibge6_list = ["410400", "330115", "330205", "353715", "355365"]
df_resultado = aggregate_morbidity_data(url, cod_ibge6_list)
df_resultado

Município  Valor_total
0  Campina Grande do Sul   33294700.7
1        Cardoso Moreira    2801938.8
2                 Italva    2395014.9
3     Pedrinhas Paulista     769964.0
4               Taquaral     382405.0

In [72]:
df_resultado.to_markdown()

'|    | Município             |      Valor_total |\n|---:|:----------------------|-----------------:|\n|  0 | Campina Grande do Sul |      3.32947e+07 |\n|  1 | Cardoso Moreira       |      2.80194e+06 |\n|  2 | Italva                |      2.39501e+06 |\n|  3 | Pedrinhas Paulista    | 769964           |\n|  4 | Taquaral              | 382405           |'